integration_tests/src/main/python/url_test.py

```diff
@@ -0,0 +1,175 @@
+    "http://foo.com/unicode_(✪)_in_parens",
+    "http://foo.com/(something)?after=parens",
+    "http://☺.damowmow.com/",
+    "http://code.google.com/events/#&product=browser",
+    "http://j.mp",
+    "ftp://foo.bar/baz",
+    r"http://foo.bar/?q=Test%20URL-encoded%20stuff",
+    "http://مثال.إختبار",
+    "http://例子.测试",
+    "http://उदाहरण.परीक्षा",
+    "http://-.~_!$&'()*+,;=:%40:80%2f::::::@example.com",
+    "http://1337.net",
+    "http://a.b-c.de",
+    "http://223.255.255.254",
+    "https://foo_bar.example.com/",
+    "http:# ",
+    "http://.",
+    "http://..",
+    "http://../",
+    "http://?",
+    "http://??",
+    "http://??/",
+    "http://#",
+    "http://##",
+    "http://##/",
+    "http://foo.bar?q=Spaces should be encoded",
+    "# ",
+    "//a",
+    "///a",
+    "/# ",
+    "http:///a",
+    "foo.com",
+    "rdar://1234",
+    "h://test",
+    "http:// shouldfail.com",
+    ":// should fail",
+    "http://foo.bar/foo(bar)baz quux",
+    "ftps://foo.bar/",
+    "http://-error-.invalid/",
+    "http://a.b--c.de/",
+    "http://-a.b.co",
+    "http://a.b-.co",
+    "http://0.0.0.0",
+    "http://10.1.1.0",
+    "http://10.1.1.255",
+    "http://224.1.1.1",
+    "http://1.1.1.1.1",
+    "http://123.123.123",
+    "http://3628126748",
+    "http://.www.foo.bar/",
+    "http://www.foo.bar./", 
+    "http://.www.foo.bar./",
+    "http://10.1.1.1",
+    "http://10.1.1.254",
+    "http://userinfo@spark.apache.org/path?query=1#Ref",
+    r"https://use%20r:pas%20s@example.com/dir%20/pa%20th.HTML?query=x%20y&q2=2#Ref%20two",
+    "http://user:pass@host",
+    "http://user:pass@host/",
+    "http://user:pass@host/?#",
+    "http://user:pass@host/file;param?query;p2",
+    "inva lid://user:pass@host/file;param?query;p2",
+    "http://[1:2:3:4:5:6:7:8]",
+    "http://[1::]",
+    "http://[1:2:3:4:5:6:7::]",
+    "http://[1::8]",
+    "http://[1:2:3:4:5:6::8]",
+    "http://[1:2:3:4:5:6::8]",
+    "http://[1::7:8]",
+    "http://[1:2:3:4:5::7:8]",
+    "http://[1:2:3:4:5::8]",
+    "http://[1::6:7:8]",
+    "http://[1:2:3:4::6:7:8]",
+    "http://[1:2:3:4::8]",
+    "http://[1::5:6:7:8]",
+    "http://[1:2:3::5:6:7:8]",
+    "http://[1:2:3::8]",
+    "http://[1::4:5:6:7:8]",
+    "http://[1:2::4:5:6:7:8]",
+    "http://[1:2::8]",
+    "http://[1::3:4:5:6:7:8]",
+    "http://[1::3:4:5:6:7:8]",
+    "http://[1::8]",
+    "http://[::2:3:4:5:6:7:8]",
+    "http://[::2:3:4:5:6:7:8]",
+    "http://[::8]",
+    "http://[::]",
+    "http://[fe80::7:8%eth0]",
+    "http://[fe80::7:8%1]",
+    "http://[::255.255.255.255]",
+    "http://[::ffff:255.255.255.255]",
+    "http://[::ffff:0:255.255.255.255]",
+    "http://[2001:db8:3:4::192.0.2.33]",
+    "http://[64:ff9b::192.0.2.33]"
+]
+
+edge_cases_gen = SetValuesGen(StringType(), edge_cases)
+
+url_gen = StringGen(url_pattern)
+    
+@pytest.mark.parametrize('data_gen', [url_gen, edge_cases_gen], ids=idfn)
+def test_parse_url_protocol(data_gen):
```

Comment:
nit: necessary to have some negative cases for protocol and host? To validate same exception behavior as Spark. Not sure the expected behavior for NaN and null string literal.

Reply:
Yes, we have many negative cases that spark returns null in tests, and we will also test against a large dataset that includes many negative cases.